Fix cuda archs

RaulPPelaez opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->
We have been seeing errors related to this feedstock not compiling for some CUDA architectures. Current build scripts try to set CUDA archs by setting the CMAKE_CUDA_ARCHITECTURES variable.
It appears that the CMake is then promptly ignoring this variable, which is instead overwritten by torch.
I find that the TORCH_CUDA_ARCH_LIST can be set instead and this is acknowledged by CMake.
Instead of relying on the current semi manual list of archs in this PR I ask torch to provide the list or archs it was itself compiled for. 
It makes sense to me that this library, which extends pytorch, simply provides the same archs at torch. However, I might be missing something, so please check.


Related issues:
https://github.com/openmm/NNPOps/issues/110
https://github.com/openmm/openmm-torch/pull/106
<!--
Please add any other relevant info below:
-->

RaulPPelaez commented:
The runners seem to be picking up the archs correctly:
```shell
-- Added CUDA NVCC flags for: -gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86
```

mikemhenry commented:
Right now github has some issues with webooks: https://www.githubstatus.com/incidents/ljhhhw2tplqp 

I will merge it once we are green

mikemhenry commented:
@RaulPPelaez Do you want to be a maintainer of the feedstock?

h-vetinari commented on recipe/build.sh:
```diff
@@ -13,39 +13,8 @@ CMAKE_FLAGS+=" -DTorch_DIR=${SP_DIR}/torch/share/cmake/Torch"
-    echo "Set up extra cmake-args: CUDA_CONFIG_ARGS=${CUDA_CONFIG_ARGS+"${CUDA_CONFIG_ARGS[@]}"}"
-
-    CMAKE_FLAGS+=" ${CUDA_CONFIG_ARGS}"
+    ARCH_LIST=$(${PYTHON} -c "import torch; print(';'.join([arch.split('_')[1][:1] + '.' + arch.split('_')[1][1:] for arch in torch._C._cuda_getArchFlags().split()]))")
```
Why use a private method rather than `torch.cuda.get_arch_list()`? AFAICT we should be able to use the output from the latter directly. Also, a bit of documentation would help.
```suggestion
    # sample output for torch.cuda.get_arch_list(): ['sm_37', 'sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86', 'sm_90']
    # we need to turn this into: "3.7;5.0;6.0;7.0;7.5;8.0;8.6;9.0"
    ARCH_LIST=$(python -c "import torch; print(';'.join([f'{y[0]}.{y[1]}' for y in [x[3:] for x in torch.cuda.get_arch_list()]]))")
```

RaulPPelaez replied:
The higher level cuda.get_arch_list is currently defined like this:
```python
def get_arch_list() -> List[str]:
    r"""Returns list CUDA architectures this library was compiled for."""
    if not is_available():
        return []
    arch_flags = torch._C._cuda_getArchFlags()
    if arch_flags is None:
        return []
    return arch_flags.split()
```
It will return nothing if there is no GPU (which I think is a mistake, tbh).
I tried that to start with, but found out it did not work. Luckily the _C call is compliant always :P

h-vetinari replied:
OK, fair enough. So then my suggestion would be
```suggestion
    # sample output for torch._C._cuda_getArchFlags(): "sm_37 sm_50 sm_60 sm_70 sm_75 sm_80 sm_86 sm_90"
    # we need to turn this into: "3.7;5.0;6.0;7.0;7.5;8.0;8.6;9.0"
    ARCH_LIST=$(python -c "import torch; print(';'.join([f'{y[0]}.{y[1]}' for y in [x[3:] for x in torch._C._cuda_getArchFlags().split()]]))")
```

RaulPPelaez replied:
What about this then?
```python
python -c "import torch; print(';'.join([f'{y[0]}.{y[1]}' for y in [x[3:] for x in torch._C._cuda_getArchFlags().split() if x.startswith('sm_')]]))"
```

h-vetinari replied:
Ah, yes, good point. I had found a sample output on the pytorch issue tracker, and it didn't have the `compute_` part.

That bit with the last arch being repeated at the end is actually relevant for the PTX thing; I'm not sure how TORCH_CUDA_ARCH does that behind the scenes (I guess it does the same). But your suggestion is fine, as long as you also update the comment above w.r.t the expected output of `_cuda_getArchFlags`.

h-vetinari reviewed (commented):
Cool if this works! Would also be happy to have you on board.

I think we could still improve the call a bit, and in any case, leave some breadcrumbs for future readers.

RaulPPelaez commented:
> Awesome! This looks way better! Where did you find:
> 
> ```
>     ARCH_LIST=$(${PYTHON} -c "import torch; print(';'.join([arch.split('_')[1][:1] + '.' + arch.split('_')[1][1:] for arch in torch._C._cuda_getArchFlags().split()]))")
>     CMAKE_FLAGS+=" -DTORCH_CUDA_ARCH_LIST=${ARCH_LIST}"
> ```
> 
> I've got a few feedstocks I would like to use this method for, how much does this depend on the version/arch of torch verses the cudatoolkit?

I am pleased to say this is of my own making hehe, feel free to take it :P
There is a higher level call `torch.cuda.get_arch_list()` but for some reason it refuses to give the information if a GPU is not present. The documentation says "return the list of archs for which torch was compiled" but this is a lie -.-
I have no idea how robust this is regarding versions of pytorch/CUDA. But even if the function signature changes I would be really surprised to see pytorch stop providing a way to produce this information.
I guess last instance you could do this:
```shell
$ cuobjdump $(find $CONDA_PREFIX -name "libtorch_cuda.so")  | grep arch | awk '{print $3}' | sort | uniq | sed 's+sm_\([0-9]\)\([0-9]\)+\1.\2+g' | tr '\n' ';'
```
which outputs the same exact info.
However, both approaches assume two things:
1. You only want to compile for the same archs as torch has
2. Your CUDA compiler understand them all (which I am not entirely sure its a given always)

RaulPPelaez commented:
> @RaulPPelaez Do you want to be a maintainer of the feedstock?

Yeah sure.

RaulPPelaez commented:
BTW! I looked around but did not really found an explanation. 
Right now one cannot do this:
```shell
$ mamba install nnpops==0.6 pytorch==1.12
```
This is because this feedstock only compiles for pytorch 2. Could you give me some insights as to why?
Is conda-forge trying to support only one pytorch version as a dependency to reduce the overall number of builds.
I also do not really get how CUDA should be handled. Right now nnpops is compiled without CUDA or with CUDA 11.2. I am aware CUDA 12 is kind of a WIP, but how about CUDA 11.8? I would expect some performance gains from it, but it seems like conda-forge really wants me to use 11.2 and thats it.
Thanks.

h-vetinari commented:
> This is because this feedstock only compiles for pytorch 2. Could you give me some insights as to why?
> Is conda-forge trying to support only one pytorch version as a dependency to reduce the overall number of builds.

We have one main version in the global [pinning](https://github.com/conda-forge/conda-forge-pinning-feedstock), because there can only be one pytorch per environment, and building different feedstocks for different pytorch versions only leads to conflicts.

The exception is during a "migration" (when feedstocks are being rebuilt for a new version), then we have a before & after version.

And if you _really_ need, you can override that mechanism by putting something as follows into your `recipe/conda_build_config.yaml`:
```
pytorch:
  - "1.13"
  - "2.0"
 ```
(+ rerender the recipe afterwards)

> I also do not really get how CUDA should be handled. Right now nnpops is compiled without CUDA or with CUDA 11.2. I am aware CUDA 12 is kind of a WIP, but how about CUDA 11.8? I would expect some performance gains from it, but it seems like conda-forge really wants me to use 11.2 and thats it.

We had several CUDA version, but recently dropped everything below 11.2. That version is special because it covers the full 11.x line after that, so a build with 11.2 can run on a 11.8 host. CUDA 12 is not WIP in conda-forge anymore, but pytorch hasn't been [rebuilt](https://github.com/conda-forge/pytorch-cpu-feedstock/pull/171) for it yet. CUDA 12 came with a gigantic overhaul of the whole infrastructure (things will be much nicer from here on out), and together with the compatibility of 11.2, there very likely won't ever be a CUDA 11.8 in conda-forge. So yes, for now it's 11.2, though any help would be appreciated in getting pytorch for CUDA 12 off the ground!

mikemhenry commented:
@conda-forge-admin, please rerender

mikemhenry commented:
I'm happy with this, but not sure where the dust settled on how we want to figure out the cuda arches, are you happy with it @h-vetinari ? I think the comment is clear at least on what it does and what to do if it breaks.

h-vetinari commented on recipe/build.sh:
```diff
@@ -10,42 +10,17 @@ function version2int { echo "$@" | awk -F. '{ printf("%d%02d\n", $1, $2); }'; }
-
-    CMAKE_FLAGS+=" ${CUDA_CONFIG_ARGS}"
+    # Output format from torch._C._cuda_getArchFlags():
+    # ['sm_37', 'sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86', 'compute_86', 'sm_90']
```
pretty sure that's wrong.
```suggestion
    # ['sm_37', 'sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86', 'compute_86', 'sm_90']
```

h-vetinari replied:
I think it's a single, space-separated string (hence the `.split()`). Please double-check.

h-vetinari commented on recipe/build.sh:
```diff
@@ -10,42 +10,17 @@ function version2int { echo "$@" | awk -F. '{ printf("%d%02d\n", $1, $2); }'; }
-    CMAKE_FLAGS+=" ${CUDA_CONFIG_ARGS}"
+    # Output format from torch._C._cuda_getArchFlags():
+    # ['sm_37', 'sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86', 'compute_86', 'sm_90']
+    # We need to turn this into: "3.7;5.0;6.0;7.0;7.5;8.0;8.6;9.0" for CMake
```
CMake has it's own format, and it's different from this one. You can see how it was constructed above, ending with.
``` 
CMAKE_CUDA_ARCHS="${CMAKE_CUDA_ARCHS};${LATEST_ARCH}-real;${LATEST_ARCH}-virtual"
```

So our adaptation is for `TORCH_CUDA_ARCH_LIST` and thus pytorch.

RaulPPelaez commented:
Ready to merge on my part!

raimis commented on recipe/build.sh:
```diff
@@ -10,42 +10,17 @@ function version2int { echo "$@" | awk -F. '{ printf("%d%02d\n", $1, $2); }'; }
+    # There is a higher level function, called torch.cuda.get_arch_list, but it returns an empty list when there is no GPU available.
+    # Should that fail, this could be used instead:
+    # $ cuobjdump $(find $CONDA_PREFIX -name "libtorch_cuda.so")  | grep arch | awk '{print $3}' | sort | uniq | sed 's+sm_\([0-9]\)\([0-9]\)+\1.\2+g' | tr '\n' ';'
+    ARCH_LIST=$(${PYTHON} -c "import torch; print(';'.join([f'{y[0]}.{y[1]}' for y in [x[3:] for x in torch._C._cuda_getArchFlags().split() if x.startswith('sm_')]]))")
```
This will work incorrectly when the compute compatibility will reach double digits. E.g. for `sm_100`, it will give `1.0` rather than `10.0`.

h-vetinari replied:
I think we'll cross that bridge when we get there. For now this is fine IMO

h-vetinari commented on recipe/build.sh:
```diff
@@ -10,42 +10,17 @@ function version2int { echo "$@" | awk -F. '{ printf("%d%02d\n", $1, $2); }'; }
-    echo "Set up extra cmake-args: CUDA_CONFIG_ARGS=${CUDA_CONFIG_ARGS+"${CUDA_CONFIG_ARGS[@]}"}"
-
-    CMAKE_FLAGS+=" ${CUDA_CONFIG_ARGS}"
+    # Output format from torch._C._cuda_getArchFlags():
+    # ['sm_35 sm_50 sm_60 sm_61 sm_70 sm_75 sm_80 sm_86 compute_86']
+    # We need to turn this into: "3.5;5.0;6.0;6.1;7.0;7.5;8.0;8.6" for pytorch in CMake
```
```suggestion
    # Output format from torch._C._cuda_getArchFlags(): 'sm_35 sm_50 sm_60 sm_61 sm_70 sm_75 sm_80 sm_86 compute_86'
    # We need to turn this into: "3.5;5.0;6.0;6.1;7.0;7.5;8.0;8.6" for TORCH_CUDA_ARCH_LIST (which overrides CMake-native option)
```

h-vetinari reviewed (approved):
LGTM, modulo a nit that I'll commit to accelerate this a bit.

Thanks for tackling this, and welcome as a maintainer on this feedstock! 🥳

mikemhenry commented:
:tada: great work everyone, thank you all so much :open_hands:
